### DISCUSSION TITLE
Unexpected Image extraction

### DISCUSSION THREAD

**Discussion by buptyyf** [pymupdf/PyMuPDF#2124] [Q&A] [answered]
_**Please provide all mandatory information!**_

## Describe the bug (mandatory)

I am working on image extraction from PDF. But pymupdf generates some png images inverted color, some png images inverted by x coordinate, and some images meet expectation.

## To Reproduce (mandatory)
This is test pdf:
[Test.pdf](https://github.com/pymupdf/PyMuPDF/files/10228440/Test.pdf)

My test code is:

```
import io
import fitz
from PIL import Image
path = 'Test.pdf'
doc = fitz.open(path, filetype="pdf")

img_index = 0
page_count = doc.page_count
if page_count:

    for page_no in range(page_count):
        blocks = doc[page_no].getText('dict')['blocks']
        for ind, block in enumerate(blocks):
            if block['type'] == 1:
                try:
                    image = Image.open(io.BytesIO(block['image']))
                    image.save(open("{}.{}".format(img_index, block['ext']), "wb"))
                    img_index += 1
                except Exception as e:
                    print(e)
```
And I also use PyMuPDF-Utilities exmaple [https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/examples/extract-images/extract.py](https://github.com/pymupdf/PyMuPDF-Utilities/blob/master/examples/extract-images/extract.py) to test extract images, which get the same result.

The result is:

1. This image should be "1961" but get this. 
<img width="603" alt="image" src="https://user-images.githubusercontent.com/13342539/207618594-cdfc27b5-b10d-4905-9a89-d761cc2a44a1.png">
<img width="480" alt="image" src="https://user-images.githubusercontent.com/13342539/207618665-ea26b29b-25af-46b1-ac2c-29baa56d80a7.png">

3. This image should be black font color and white background color, but get the inverted color 
![img17230](https://user-images.githubusercontent.com/13342539/207617345-df97c1d3-1c11-4591-967e-88534b89b4f0.png)
<img width="475" alt="image" src="https://user-images.githubusercontent.com/13342539/207617672-9181c186-88e8-4c01-aa55-b946792cb109.png">

4. This image behavior meet my expectation.

![img17240](https://user-images.githubusercontent.com/13342539/207617812-e775e325-86c7-44dc-9b9b-7f75203fa06d.png)
<img width="484" alt="image" src="https://user-images.githubusercontent.com/13342539/207617865-d26ec3a9-24cf-4b80-a897-79b5368cc297.png">

**Comment by JorjMcKie** [comment]
As is mentioned in the documentation:
If you extract images via `Page.get_text()`, then you will **_not get transparency_** information. Only base images are extracted.
Whatever your expectations are: don't be disappointed if extracting this way.

But you actually are lucky: your 2100+ images seem to have no transparency and with only 6 exceptions they **_all are extracted correctly!_**
If they seem to be flipped or rotated, then this means, they are stored that way - and the page creator knew what he must do to show them more nicely.
You can find out what he did by looking at the `"tranform"` matrix in each image block. The documentation will tell you how to interpret that matrix.
Example page 284, image block 0:
```python
>>> blocks[0].keys()
dict_keys(['number', 'type', 'bbox', 'width', 'height', 'ext', 'colorspace', 'xres', 'yres', 'bpc', 'transform', 'size', 'image'])
>>> blocks[0]["bpc"]
1
>>> blocks[0]["transform"]
(3.0, 0.0, -0.0, -1.8461999893188477, 294.5830078125, 414.6283874511719)
```
This shows:
* the transform `matrix.d` is < 0, therefore an up-down flip
* the bpc = 1, therefore this image is black&white. It is one of the 6 mentioned exceptions. To store it in a more accustomed way, perform a color inversion like that:
```python
pix = fitz.Pixmap(blocks[0]["image"])
pix.invert_irect()
pix.save("inverted.png")
```

**Reply by buptyyf** [reply]
Thanks. But this method don't solve my problems. I also have some problems as follows:

1. up-down flip problem.
"the transform matrix.d is < 0, therefore an up-down flip"
I want to flip it back, making it likes normal. pix.invert_irect() just inverted color. What should I do? 

2. get_text() can't extract image transparency info.
I learn from [you answer](https://github.com/pymupdf/PyMuPDF/issues/1919) use doc.extract_image api to extract image with transparency info. But it's not working for these images. doc.extract_image() "image" property value is same as page.get_text() block "image" property.
<img width="475" alt="image" src="https://user-images.githubusercontent.com/13342539/207617672-9181c186-88e8-4c01-aa55-b946792cb109.png" style="max-width: 100%;">  My understanding is that, when I extract the transparency of image, these images will display normally. I don't know whether my understanding is right.

3. How to one on one map image block from get_text() with page.get_images() list item?
 Because page.get_images() list item has xref, but image block from get_text() don't have xref property.

**Comment by JorjMcKie** [comment]
You may have questions, so I am converting this to a "Discussions" item.

**Comment by JorjMcKie** [comment, marked as answer]
> How to one on one map image block from get_text() with page.get_images() list item?
Because page.get_images() list item has xref, but image block from get_text() don't have xref property.

A 1:1 correspondence between `page.get_images()` and the image blocks in `page.get_text()` **_cannot exist / be guaranteed_**, see the documentation [here](https://pymupdf.readthedocs.io/en/latest/textpage.html#block-dictionaries).

To recover images in the same way as a page shows them, you must investigate the `transform` matrix and use a package like Pillow to back-transform the extracted image accordingly.
I have once written a little function `matprop`, that does this investigation. Here is a ZIP of it:
[matrix_property.zip](https://github.com/pymupdf/PyMuPDF/files/10236026/matrix_property.zip)
Lets look at page 8 of your test file and see how to use it:
```python
import fitz
from PIL import Image
import io
from matrix_property import matprop
doc=fitz.open("Test.pdf")
page = doc[8]
imgb = [b for b in page.get_text("dict")["blocks"] if b["type==1]]
for i, b in enumerate(imgb):
    image0 = Image.open(io.BytesIO(b["image"]))
    mat = fitz.Matrix(b["transform"])
    if matprop(mat)[0] == 2:  # this is an up-down flip!
        image1 = image0.transpose(Image.Transpose.FLIP_TOP_BOTTOM)
        image1.save(f"img{i}.{b"ext"})
```
This will produce the following images:
![grafik](https://user-images.githubusercontent.com/8290722/207835812-57a405f2-b83f-4a2c-ba06-43edad3e5336.png)
Depending on the returns of function `matprop`, select the right `Image.Transpose` function to achieve your result.

**Reply by buptyyf** [reply]
Thanks a lot! But I can't find the answer of No.2 problem.
>get_text() can't extract image transparency info.
I learn from https://github.com/pymupdf/PyMuPDF/issues/1919 use doc.extract_image api to extract image with transparency info. But it's not working for these images. doc.extract_image() "image" property value is same as page.get_text() block "image" property.
<image> My understanding is that, when I extract the transparency of image, these images will display normally. I don't know whether my understanding is right.

**Reply by JorjMcKie** [reply]
> My understanding is that, when I extract the transparency of image, these images will display normally. I don't know whether my understanding is right.

No, this is a misconception. All the images in your test file are intransparent, meaning they do not have an alpha channel.
Therefore your extraction via get_text() was successful.
If you use the PDF-specific approach (remember: get_text works for all supported document types - not only PDF) `Page.get_images()`, then you will be given the image's xref in position 0 of each item **_and_** the xref of the transparency mask in position 1. If this value is > 0, then the image given in position 0 must be modified:
* add an alpha channel
* put the mask bytes into the new alpha channel
* then proceed with the resulting pixmap

To find out the transform matrix in this situation again, use `Page.get_image_rects(xref, transform=True)`, and proceed like above.

**Reply by buptyyf** [reply]
> All the images in your test file are intransparent, meaning they do not have an alpha channel. Therefore your extraction via get_text() was successful.
 
I don't know why these images are black background color when they are extracted by get_text(). But show white background color in pdf file. The smask of all these images is 0.

I know I can use pix.invert_irect() to handle these six images. But how can I detect which image should be inverted in other pdf files? Does it mean when bpc is 1, I should inverted that?

**Reply by JorjMcKie** [reply]
bpc=1 is the only case where this is necessary.
Otherwise it is a bug, which you should report please.

**Reply by buptyyf** [reply]
Both of this image <img width="480" alt="image" src="https://user-images.githubusercontent.com/13342539/207618665-ea26b29b-25af-46b1-ac2c-29baa56d80a7.png" style="max-width: 100%;"> and this image<img src="https://user-images.githubusercontent.com/13342539/207617345-df97c1d3-1c11-4591-967e-88534b89b4f0.png" alt="img17230" style="max-width: 100%;"> have the same property bpc = 1. But only the second image should be inverted color. How can I detect which image should be inverted?

**Comment by buptyyf** [comment]
Another question about extract image.

[CB_Aj05vi5wT4PZ6bp6cf.pdf](https://github.com/pymupdf/PyMuPDF/files/10242641/CB_Aj05vi5wT4PZ6bp6cf.pdf)

I extract the first image in page 0 by two ways.

1. I use get_text() api extract image content directly. This image change to black background color, but it's white background in pdf.
 
<img width="411" alt="image" src="https://user-images.githubusercontent.com/13342539/208021396-5ab7885b-4373-4f73-b1eb-ea94496eb5b3.png">

2. Then I checked this image as follows:
```
>>> doc.extract_image(14)
{'ext': 'jpeg', 'smask': 15, 'width': 1244, 'height': 514, 'colorspace': 3, 'bpc': 8, 'xres': 96, 'yres': 96, 'cs-name': 'DeviceRGB', 'image': b'\xff\xd8\xff\xe0\x00\x10JFIF\x00\x01\x01\x01\x00`\x00`\x00\x00...(\xa2\x80\n(\xa2\x80?\xff\xd9'}
>>> doc.extract_image(15)
{'ext': 'png', 'smask': 0, 'width': 1244, 'height': 514, 'colorspace': 1, 'bpc': 8, 'xres': 96, 'yres': 96, 'cs-name': 'DeviceGray', 'image': b'\x89PNG\r\n\x1a\n\x00\x00\x00\rIHDR\x00\x00\x04\xdc...\x00\x00IEND\xaeB`\x82'}
>>> pix14 = fitz.Pixmap(doc, 14)

>>>pix15 = fitz.Pixmap(doc, 15)

>>>pix = fitz.Pixmap(pix14, pix15)
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/Users/yifei/.pyenv/versions/3.10.7/envs/wrpdf_proj/lib/python3.10/site-packages/fitz/fitz.py", line 6820, in __init__
    _fitz.Pixmap_swiginit(self, _fitz.new_Pixmap(*args))
RuntimeError: color pixmap must not have an alpha channel
```

I use for [reference](https://github.com/pymupdf/PyMuPDF/issues/1406) your method, but get some errors.

In addition, the bpc of this image is 8, not 1,  so I don't use pix.invert_irect() api to invert it.

**Comment by JorjMcKie** [comment]
This is a base image with an SMask (= alpha channel) stored. So the `get_text()` method will fail for it.
You then correctly extracted both, base image and SMask. But the PDF creator stored the base image with a (presumably useless) alpha channel - that's what the error message complains about.

You can remove the alpha channel from the base image before you add the SMask pixmap:
```python
if pix14.alpha:  # check if alpha channel there
    pix14 = fitz.Pixmap(pix14, 0)  # remove alpha channel
pix = fitz.Pixmap(pix14, pix15)
```

**Reply by buptyyf** [reply]
Almighty! You can solve everything. Thanks a lot! 
And this [question](https://github.com/pymupdf/PyMuPDF/discussions/2124#discussioncomment-4415910) at above also need your help. All of these images' smask are 0, but some need inverted and others don't need.

**Reply by JorjMcKie** [reply]
smask = 0 does not determine whether or not to invert the pixmap - the existence of an smask > 0 only says we need an alpha channel.
But if img["bpc"] = 1 (and especially if also img["cs-name"] = "None"), the inversion is (probably - there may be issues) required.

**Reply by buptyyf** [reply]
Okay!
img["cs-name"] = "None" is a key point I missed.

Thanks again!!!
